src/lune/builtins/datetime/date_time.rs

```diff
@@ -0,0 +1,243 @@
+use crate::lune::builtins::datetime::builder::DateTimeBuilder;
+use chrono::prelude::*;
+use chrono::DateTime as ChronoDateTime;
+
+/// Possible types of timestamps accepted by `DateTime`.
+pub enum TimestampType {
+    Seconds,
+    Millis,
+}
+
+/// General timezone types accepted by `DateTime` methods.
+pub enum Timezone {
+    Utc,
+    Local,
+}
+
+#[derive(Clone)]
+pub struct DateTime {
+    /// The number of **seconds** since January 1st, 1970
+    /// at 00:00 UTC (the Unix epoch). Range is
+    /// -17,987,443,200 to 253,402,300,799, approximately
+    /// years 1400–9999.
+    pub unix_timestamp: i64,
+
+    /// The number of **milliseconds* since January 1st, 1970
+    /// at 00:00 UTC (the Unix epoch). Range is -17,987,443,200,000
+    /// to 253,402,300,799,999, approximately years 1400–9999.
+    pub unix_timestamp_millis: i64,
+}
+
+impl DateTime {
+    /// Returns a `DateTime` representing the current moment in time.
+    pub fn now() -> Self {
+        let time = Utc::now();
+
+        Self {
+            unix_timestamp: time.timestamp(),
+            unix_timestamp_millis: time.timestamp_millis(),
+        }
+    }
+
+    /// Returns a new `DateTime` object from the given unix timestamp, in either seconds on
+    /// milliseconds. In case of failure, defaults to the  (seconds or
+    /// milliseconds) since January 1st, 1970 at 00:00 (UTC).
+    pub fn from_unix_timestamp(timestamp_kind: TimestampType, unix_timestamp: i64) -> Self {
+        let time_chrono = match timestamp_kind {
+            TimestampType::Seconds => NaiveDateTime::from_timestamp_opt(unix_timestamp, 0),
+            TimestampType::Millis => NaiveDateTime::from_timestamp_millis(unix_timestamp),
+        };
+
+        if let Some(time) = time_chrono {
+            Self {
+                unix_timestamp: time.timestamp(),
+                unix_timestamp_millis: time.timestamp_millis(),
+            }
+        } else {
+            Self::now()
+        }
+    }
+
+    /// Returns a new `DateTime` using the given units from a UTC time. The
+    /// values accepted are similar to those found in the time value table
+    /// returned by `to_universal_time`.
+    ///
+    /// - Date units (year, month, day) that produce an invalid date will raise an error. For example, January 32nd or February 29th on a non-leap year.
+    /// - Time units (hour, minute, second, millisecond) that are outside their normal range are valid. For example, 90 minutes will cause the hour to roll over by 1; -10 seconds will cause the minute value to roll back by 1.
+    /// - Non-integer values are rounded down. For example, providing 2.5 hours will be equivalent to providing 2 hours, not 2 hours 30 minutes.
+    /// - Omitted values are assumed to be their lowest value in their normal range, except for year which defaults to 1970.
+    pub fn from_universal_time(date_time: Option<DateTimeBuilder>) -> Self {
```

Comment:
Based on the spec that I'm going off of, from_universal_time can either take an argument with the date and time defined, or return the current moment in time. This is more or less for semantics' sake.

Reply:
That's fine, but someone else reading this, or needing to use it in the future, wouldn't know that. Mistakes happen, people forget or don't always read entire function bodies to figure out their semantics, it's better to make things overly obvious than not.

Some examples from a consumer perspective, constructor that doesn't take an option:

```rust
let dt = match dt_builder {
    Some(builder) => DateTime::from_universal_time(builder),
    None => DateTime::now(),
};
```

Takes an option:

```rust
// If you already have an option
let dt = DateTime::from_universal_time(builder_opt);
// If you don't
let dt = DateTime::from_universal_time(Some(builder));
let dt = DateTime::from_universal_time(None);
```

It's really not obvious what's going on in the second code block and that it will fallback to the `now` default

Reply:
That looks great! I thought you meant removing the usage of options altogether.

Reply:
I've done this for from_universal_time, however I cannot do this for from_local_time since DateTime::now() returns a datetime as per the universal time.